File: padlock/src/main/java/com/pyamsoft/padlock/pin/PinEntryTextFragment.java
/*
 * Copyright 2016 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.padlock.pin;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import com.pyamsoft.padlock.Injector;
import com.pyamsoft.padlock.PadLock;
import com.pyamsoft.padlock.R;
import com.pyamsoft.padlock.databinding.FragmentPinEntryTextBinding;
import com.pyamsoft.pydroid.bus.EventBus;
import com.pyamsoft.pydroid.ui.loader.DrawableHelper;
import com.pyamsoft.pydroid.ui.loader.DrawableLoader;
import javax.inject.Inject;
import timber.log.Timber;

public class PinEntryTextFragment extends PinEntryBaseFragment {

  @NonNull static final String TAG = "PinEntryTextFragment";
  @NonNull private static final String CODE_DISPLAY = "CODE_DISPLAY";
  @NonNull private static final String CODE_REENTRY_DISPLAY = "CODE_REENTRY_DISPLAY";
  @NonNull private static final String HINT_DISPLAY = "HINT_DISPLAY";
  @SuppressWarnings("WeakerAccess") @Inject PinEntryPresenter presenter;
  EditText pinEntryText;
  EditText pinHintText;
  @SuppressWarnings("WeakerAccess") InputMethodManager imm;
  FragmentPinEntryTextBinding binding;
  private EditText pinReentryText;
  @NonNull final PinEntryPresenter.SubmitCallback submitCallback =
      new PinEntryPresenter.SubmitCallback() {

        @Override public void onSubmitSuccess(boolean creating) {
          clearDisplay();
          if (creating) {
            EventBus.get().publish(CreatePinEvent.create(true));
          } else {
            EventBus.get().publish(ClearPinEvent.create(true));
          }
          dismissParent();
        }

        @Override public void onSubmitFailure(boolean creating) {
          clearDisplay();
          if (creating) {
            EventBus.get().publish(CreatePinEvent.create(false));
          } else {
            EventBus.get().publish(ClearPinEvent.create(false));
          }
          dismissParent();
        }

        @Override public void onSubmitError() {
          clearDisplay();
          dismissParent();
        }
      };
  @NonNull private DrawableLoader.Loaded goTask = DrawableLoader.empty();

  @Override public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    Injector.get().provideComponent().plusPinEntryComponent().inject(this);
  }

  @Nullable @Override
  public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
      @Nullable Bundle savedInstanceState) {
    binding = FragmentPinEntryTextBinding.inflate(inflater, container, false);
    return binding.getRoot();
  }

  @Override public void onDestroyView() {
    super.onDestroyView();
    goTask = DrawableHelper.unload(goTask);
    imm.toggleSoftInputFromWindow(getActivity().getWindow().getDecorView().getWindowToken(), 0, 0);
    binding.unbind();
  }

  @Override public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);
    // Resolve TextInputLayout edit texts
    pinEntryText = binding.pinEntryCode.getEditText();
    if (pinEntryText == null) {
      throw new NullPointerException("No item entry edit text");
    }

    pinReentryText = binding.pinReentryCode.getEditText();
    if (pinReentryText == null) {
      throw new NullPointerException("No item re-entry edit text");
    }

    pinHintText = binding.pinHint.getEditText();
    if (pinHintText == null) {
      throw new NullPointerException("No item hint edit text");
    }

    // Force the keyboard
    imm = (InputMethodManager) getContext().getApplicationContext()
        .getSystemService(Context.INPUT_METHOD_SERVICE);
    imm.toggleSoftInput(InputMethodManager.SHOW_FORCED, InputMethodManager.HIDE_IMPLICIT_ONLY);

    clearDisplay();
    setupGoArrow(submitCallback);

    if (savedInstanceState != null) {
      onRestoreInstanceState(savedInstanceState);
    }
  }

  @Override public void onStart() {
    super.onStart();
    presenter.checkMasterPinPresent(new PinEntryPresenter.MasterPinStatusCallback() {

      private void setupSubmissionView(@NonNull EditText view,
          @NonNull PinEntryPresenter.SubmitCallback submitCallback) {
        view.setOnEditorActionListener((textView, actionId, keyEvent) -> {
          if (keyEvent == null) {
            Timber.e("KeyEvent was not caused by keypress");
            return false;
          }

          if (keyEvent.getAction() == KeyEvent.ACTION_DOWN && actionId == EditorInfo.IME_NULL) {
            Timber.d("KeyEvent is Enter pressed");
            presenter.submit(getCurrentAttempt(), getCurrentReentry(), getCurrentHint(),
                submitCallback);
            return true;
          }

          Timber.d("Do not handle key event");
          return false;
        });
      }

      @Override public void onMasterPinMissing() {
        Timber.d("No active master, show extra views");
        binding.pinReentryCode.setVisibility(View.VISIBLE);
        binding.pinHint.setVisibility(View.VISIBLE);
        setupSubmissionView(pinHintText, submitCallback);
      }

      @Override public void onMasterPinPresent() {
        Timber.d("Active master, hide extra views");
        binding.pinReentryCode.setVisibility(View.GONE);
        binding.pinHint.setVisibility(View.GONE);
        setupSubmissionView(pinEntryText, submitCallback);
      }
    });
  }

  @Override public void onStop() {
    super.onStop();
    presenter.stop();
  }

  @Override public void onDestroy() {
    super.onDestroy();
    presenter.destroy();
    PadLock.getRefWatcher(this).watch(this);
  }

  private void setupGoArrow(@NonNull PinEntryPresenter.SubmitCallback submitCallback) {
    binding.pinImageGo.setOnClickListener(view -> {
      presenter.submit(getCurrentAttempt(), getCurrentReentry(), getCurrentHint(), submitCallback);
      imm.toggleSoftInputFromWindow(getActivity().getWindow().getDecorView().getWindowToken(), 0,
          0);
    });

    // Force keyboard focus
    pinEntryText.requestFocus();

    goTask = DrawableHelper.unload(goTask);
    goTask = DrawableLoader.load(R.drawable.ic_arrow_forward_24dp)
        .tint(R.color.orangeA200)
        .into(binding.pinImageGo);
  }

  private void onRestoreInstanceState(@NonNull Bundle savedInstanceState) {
    Timber.d("onRestoreInstanceState");
    final String attempt = savedInstanceState.getString(CODE_DISPLAY, null);
    final String reentry = savedInstanceState.getString(CODE_REENTRY_DISPLAY, null);
    final String hint = savedInstanceState.getString(HINT_DISPLAY, null);
    if (attempt == null || reentry == null || hint == null) {
      Timber.d("Empty attempt");
      clearDisplay();
    } else {
      Timber.d("Set attempt %s", attempt);
      pinEntryText.setText(attempt);
      Timber.d("Set reentry %s", reentry);
      pinReentryText.setText(reentry);
      Timber.d("Set hint %s", hint);
      pinHintText.setText(hint);
    }
  }

  @Override public void onSaveInstanceState(@NonNull Bundle outState) {
    Timber.d("onSaveInstanceState");
    outState.putString(CODE_DISPLAY, getCurrentAttempt());
    outState.putString(CODE_REENTRY_DISPLAY, getCurrentReentry());
    outState.putString(HINT_DISPLAY, getCurrentHint());
    super.onSaveInstanceState(outState);
  }

  /**
   * Clear the display of all text entry fields
   */
  void clearDisplay() {
    pinEntryText.setText("");
    pinReentryText.setText("");
    pinHintText.setText("");
  }

  @SuppressWarnings("WeakerAccess") @CheckResult @NonNull String getCurrentAttempt() {
    return pinEntryText.getText().toString();
  }

  @SuppressWarnings("WeakerAccess") @CheckResult @NonNull String getCurrentReentry() {
    return pinReentryText.getText().toString();
  }

  @SuppressWarnings("WeakerAccess") @CheckResult @NonNull String getCurrentHint() {
    return pinHintText.getText().toString();
  }
}

File: padlock/src/main/java/com/pyamsoft/padlock/pin/PinEntryTextFragment.kt
/*
 * Copyright (C) 2018 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.padlock.pin

import android.content.Context
import android.os.Bundle
import android.support.annotation.CheckResult
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import com.pyamsoft.padlock.Injector
import com.pyamsoft.padlock.PadLockComponent
import com.pyamsoft.padlock.databinding.FragmentPinEntryTextBinding
import com.pyamsoft.pydroid.ui.util.Snackbreak
import com.pyamsoft.pydroid.ui.util.Snackbreak.ErrorDetail
import timber.log.Timber
import javax.inject.Inject

class PinEntryTextFragment : PinEntryBaseFragment(), PinEntryPresenter.View {

  @field:Inject
  internal lateinit var presenter: PinEntryPresenter
  private lateinit var imm: InputMethodManager
  private lateinit var binding: FragmentPinEntryTextBinding
  private var pinReentryText: EditText? = null
  private var pinEntryText: EditText? = null
  private var pinHintText: EditText? = null

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    Injector.obtain<PadLockComponent>(requireContext().applicationContext)
        .inject(this)
  }

  override fun onCreateView(
    inflater: LayoutInflater,
    container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View? {
    binding = FragmentPinEntryTextBinding.inflate(inflater, container, false)
    return binding.root
  }

  override fun onDestroyView() {
    super.onDestroyView()
    activity?.let {
      imm.toggleSoftInputFromWindow(it.window.decorView.windowToken, 0, 0)
    }
    binding.unbind()
  }

  override fun onViewCreated(
    view: View,
    savedInstanceState: Bundle?
  ) {
    super.onViewCreated(view, savedInstanceState)
    // Resolve TextInputLayout edit texts
    pinEntryText = binding.pinEntryCode.editText!!
    pinReentryText = binding.pinReentryCode.editText!!
    pinHintText = binding.pinHint.editText!!

    // Force the keyboard
    imm = requireContext().getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
    imm.toggleSoftInput(InputMethodManager.SHOW_FORCED, InputMethodManager.HIDE_IMPLICIT_ONLY)

    clearDisplay()
    setupGoArrow()

    if (savedInstanceState != null) {
      onRestoreInstanceState(savedInstanceState)
    }

    presenter.bind(viewLifecycle, this)
  }

  private fun setupSubmissionView(view: EditText) {
    view.setOnEditorActionListener { _, actionId, keyEvent ->
      if (keyEvent == null) {
        Timber.e("KeyEvent was not caused by keypress")
        return@setOnEditorActionListener false
      }

      if (keyEvent.action == KeyEvent.ACTION_DOWN && actionId == EditorInfo.IME_NULL) {
        Timber.d("KeyEvent is Enter pressed")
        submitPin()
        return@setOnEditorActionListener true
      }

      Timber.d("Do not handle key event")
      return@setOnEditorActionListener false
    }
  }

  override fun onPinSubmitCreateSuccess() {
    Timber.d("Create success")
  }

  override fun onPinSubmitCreateFailure() {
    Timber.d("Create failure")
  }

  override fun onPinSubmitClearSuccess() {
    Timber.d("Clear success")
  }

  override fun onPinSubmitClearFailure() {
    Timber.d("Clear failure")
  }

  override fun onPinSubmitError(throwable: Throwable) {
    Snackbreak.short(
        requireActivity(), binding.root,
        ErrorDetail("PIN submission error", throwable.localizedMessage)
    )
  }

  override fun onPinSubmitComplete() {
    clearDisplay()
    dismissParent()
  }

  private fun submitPin() {
    // Hint is blank for PIN code
    presenter.submit(getCurrentAttempt(), getCurrentReentry(), getCurrentHint())
  }

  override fun onMasterPinMissing() {
    Timber.d("No active master, show extra views")
    binding.pinReentryCode.visibility = View.VISIBLE
    binding.pinHint.visibility = View.VISIBLE
    val obj = pinHintText
    if (obj != null) {
      setupSubmissionView(obj)
    }
  }

  override fun onMasterPinPresent() {
    Timber.d("Active master, hide extra views")
    binding.pinReentryCode.visibility = View.GONE
    binding.pinHint.visibility = View.GONE
    val obj = pinEntryText
    if (obj != null) {
      setupSubmissionView(obj)
    }
  }

  override fun onSubmitPressed() {
    submitPin()
  }

  private fun setupGoArrow() {
    // Force keyboard focus
    pinEntryText?.requestFocus()
  }

  private fun onRestoreInstanceState(savedInstanceState: Bundle) {
    Timber.d("onRestoreInstanceState")
    val attempt = savedInstanceState.getString(CODE_DISPLAY, null)
    val reentry = savedInstanceState.getString(CODE_REENTRY_DISPLAY, null)
    val hint = savedInstanceState.getString(HINT_DISPLAY, null)
    if (attempt == null || reentry == null || hint == null) {
      Timber.d("Empty attempt")
      clearDisplay()
    } else {
      Timber.d("Set attempt %s", attempt)
      pinEntryText?.setText(attempt)
      Timber.d("Set reentry %s", reentry)
      pinReentryText?.setText(reentry)
      Timber.d("Set hint %s", hint)
      pinHintText?.setText(hint)
    }
  }

  override fun onSaveInstanceState(outState: Bundle) {
    Timber.d("onSaveInstanceState")
    outState.putString(CODE_DISPLAY, getCurrentAttempt())
    outState.putString(CODE_REENTRY_DISPLAY, getCurrentReentry())
    outState.putString(HINT_DISPLAY, getCurrentHint())
    super.onSaveInstanceState(outState)
  }

  /**
   * Clear the display of all text entry fields
   */
  private fun clearDisplay() {
    pinEntryText?.setText("")
    pinReentryText?.setText("")
    pinHintText?.setText("")
  }

  @CheckResult
  private fun getCurrentAttempt(): String = pinEntryText?.text?.toString() ?: ""

  @CheckResult
  private fun getCurrentReentry(): String = pinReentryText?.text?.toString() ?: ""

  @CheckResult
  private fun getCurrentHint(): String = pinHintText?.text?.toString() ?: ""

  companion object {

    internal const val TAG = "PinEntryTextFragment"
    private const val CODE_DISPLAY = "CODE_DISPLAY"
    private const val CODE_REENTRY_DISPLAY = "CODE_REENTRY_DISPLAY"
    private const val HINT_DISPLAY = "HINT_DISPLAY"
  }
}
